api/main.go

```diff
@@ -41,60 +43,97 @@ func authGuard(core *Core) fiber.Handler {
-				"error": "Missing Authorization header",
+			return ctx.Status(fiber.StatusUnauthorized).JSON(ErrorResponse{
+				http.StatusUnauthorized,
+				"Authorization header is required",
+				nil,
 			})
 		}
 
 		const prefix = "Basic "
 		if !strings.HasPrefix(authHeader, prefix) {
-			return ctx.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
-				"error": "Invalid Authorization header",
+			return ctx.Status(fiber.StatusUnauthorized).JSON(ErrorResponse{
+				http.StatusUnauthorized,
+				"Invalid authorisation header",
+				nil,
 			})
 		}
 
 		encodedCredentials := authHeader[len(prefix):]
 		decoded, err := base64.StdEncoding.DecodeString(encodedCredentials)
 		if err != nil {
-			return ctx.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
-				"error": "Invalid base64 encoding",
-			})
+			return ctx.Status(fiber.StatusUnauthorized).JSON(
+				ErrorResponse{
+					http.StatusUnauthorized,
+					"Invalid base64 encoding",
+					nil,
+				})
 		}
 
 		credentials := strings.SplitN(string(decoded), ":", 2)
 		if len(credentials) != 2 {
-			return ctx.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
-				"error": "Invalid Authorization value",
-			})
+			return ctx.Status(fiber.StatusUnauthorized).JSON(
+				ErrorResponse{
+					http.StatusUnauthorized,
+					"Invalid Authorization value",
+					nil,
+				})
 		}
 
 		username := credentials[0]
 		password := credentials[1]
 
 		principal, err := core.vault.Login(ctx.Context(), username, password)
 		if err != nil {
-			return ctx.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
-				"error": "Invalid username or password",
-			})
+			return ctx.Status(fiber.StatusUnauthorized).JSON(
+				ErrorResponse{
+					http.StatusUnauthorized,
+					"Invalid username or password",
+					nil,
+				})
 		}
 
-		// Set principal in context
 		ctx.Locals(PRINCIPAL_CONTEXT_KEY, principal)
-		// Continue stack
-		_ = ctx.Next()
+		err = ctx.Next()
+
+		return err
+	}
+}
+
+func customErrorHandler(ctx *fiber.Ctx, err error) error {
+	var e *fiber.Error
+	if errors.As(err, &e) {
+		code := e.Code
+		return ctx.Status(code).SendString(err.Error())
+	}
 
-		return nil
+	// Handle custom errors from the vault package
+	var ve *_vault.ValueError
+	var fe *_vault.ForbiddenError
+	var ne *_vault.NotFoundError
+	switch {
+	case errors.As(err, &ve):
+		return ctx.Status(http.StatusBadRequest).JSON(ErrorResponse{http.StatusBadRequest, ve.Error(), nil})
+	case errors.As(err, &fe):
+		return ctx.Status(http.StatusForbidden).JSON(ErrorResponse{http.StatusForbidden, fe.Error(), nil})
+	case errors.As(err, &ne):
+		return ctx.Status(http.StatusNotFound).JSON(ErrorResponse{http.StatusNotFound, ne.Error(), nil})
+	default:
+		// Handle other types of errors by returning a generic 500 - this should remain obscure as it can leak information
+		return ctx.Status(http.StatusInternalServerError).JSON(ErrorResponse{http.StatusInternalServerError, "Internal Server Error", nil})
 	}
```

Comment:
Can be simplified with a switch on the err type:
`switch err.(type)`
```suggestion
	// Handle custom errors from the vault package
	switch err.(type) {
	case *_vault.ValueError:
		return ctx.Status(http.StatusBadRequest).JSON(ErrorResponse{http.StatusBadRequest, ve.Error(), nil})
	case *_vault.ForbiddenError:
		return ctx.Status(http.StatusForbidden).JSON(ErrorResponse{http.StatusForbidden, fe.Error(), nil})
	case *_vault.NotFoundError:
		return ctx.Status(http.StatusNotFound).JSON(ErrorResponse{http.StatusNotFound, ne.Error(), nil})
	default:
		// Handle other types of errors by returning a generic 500 - this should remain obscure as it can leak information
		return ctx.Status(http.StatusInternalServerError).JSON(ErrorResponse{http.StatusInternalServerError, "Internal Server Error", nil})
	}
```

Reply:
Not sure but all tests are passing fine for me:
```go
	// Handle custom errors from the vault package
	switch err.(type) {
	case *_vault.ValueError:
		return ctx.Status(http.StatusBadRequest).JSON(ErrorResponse{err.Error(), nil})
	case *_vault.ForbiddenError:
		return ctx.Status(http.StatusForbidden).JSON(ErrorResponse{err.Error(), nil})
	case *_vault.NotFoundError:
		return ctx.Status(http.StatusNotFound).JSON(ErrorResponse{err.Error(), nil})
	case *AuthError:
		return ctx.Status(http.StatusUnauthorized).JSON(ErrorResponse{err.Error(), nil})
	default:
		// Handle other types of errors by returning a generic 500 - this should remain obscure as it can leak information
		return ctx.Status(http.StatusInternalServerError).JSON(ErrorResponse{"Internal Server Error", nil})
	}
```